### ISSUE TITLE
Jolly Juniper Mouse - Distributor.sol#_executeClaim - Risk of silent overflow, which could potentially lead to a loss of funds for the beneficiaries.

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2023-06-tokensoft-judging#70]
Jolly Juniper Mouse

medium

# Distributor.sol#_executeClaim - Risk of silent overflow, which could potentially lead to a loss of funds for the beneficiaries.

## Summary

In `_executeClaim()` there is a risk of a silent overflow, which could potentially lead to a loss of funds for the beneficiaries.

## Vulnerability Detail

The `getClaimableAmount()` function returns a `uint256`:

```solidity
function getClaimableAmount(address beneficiary) public view virtual returns (uint256) {
```

It's used in the `_executeClaim()` function where is downcasted into a `uint120`:

```solidity
uint120 claimableAmount = uint120(getClaimableAmount(beneficiary));
```

After that, there's a require statement that only checks if `claimableAmount` is greater than 0 and not if it's less than or equal to type(uint120).max, which is the correct condition in this case:

```solidity
require(claimableAmount > 0, 'Distributor: no more tokens claimable right now');
```

However, there's a issue if `claimableAmount` becomes greater than type(uint120).max + 2 (if it becomes type(uint120).max + 1, the result will be 0, and it will revert because of the require). In this scenario, `claimableAmount` will silently overflow, potentially causing a loss of funds for the beneficiaries, as they might claim fewer tokens than they are entitled to.

The following PoC code confirms my finding, it can be tested on Remix or another IDE:

```solidity
contract Test {
    function _executeClaimTest (uint256 _beneficiary) public pure returns (uint256) {
        uint120 claimableAmount = uint120(_beneficiary);
        require(claimableAmount > 0,"Distributor: no more tokens claimable right now");
        return claimableAmount;
    }
}
```

1. We call the `_executeClaimTest()` function and set `_beneficiary` as 1329227995784915872903807060280344577( type(uint120).max + 2).
2. `_beneficiary` downcasts from uint256 to uint120, it silently overflows, and becomes 1.
3. `claimableAmount` becomes `uint120(_beneficiary)` which is 1.
4. It passes the require.
5. As a result, the beneficiary might claim only 1 token even though they were entitled to claim more.

## Impact

There'll be a loss of funds for the beneficiaries, because they might not be able to claim the correct amount of tokens they were entitled to.

## Code Snippet

[Distributor.sol#_executeClaim()](https://github.com/sherlock-audit/2023-06-tokensoft/blob/main/contracts/contracts/claim/abstract/Distributor.sol#L66C2-L85C4)

## Tool used

Manual Review

## Recommendation

To fix this , you should update the require statement to also check if `claimableAmount` is less than or equal to type(uint120).max:

```solidity
    require(claimableAmount <= type(uint120).max, 'Distributor: claimableAmount is greater than type(uint120).max');
```

This way, the function will revert if claimableAmount is greater than type(uint120).max, preventing any potential loss of tokens due to silent overflow.

**Comment by cholakovvv** [comment]
Escalate
This issue must be a duplicat of #165, #15 and #152. In the final part of the vulnerability detail in these reports, they provide an example of downcasting in the _executeClaim() function. This is the same issue that I am reporting, and I have also provided a PoC on Remix that prove it. Also it must be upgraded to High, because the others are.

**Comment by sherlock-admin2** [comment]
> Escalate
> This issue must be a duplicat of #165, #15 and #152. In the final part of the vulnerability detail in these reports, they provide an example of downcasting in the _executeClaim() function. This is the same issue that I am reporting, and I have also provided a PoC on Remix that prove it. Also it must be upgraded to High, because the others are.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by Shogoki** [comment]
I do not think it is a duplicate, as the other reports are about the overflow in `initializeDistributionRecord`. 
This report is about an overflow if the claimableAmount is bigger than uint120 max which i think, is not possible.
claimable Amount is calculated as follows:

```solidity
    uint256 claimable = (record.total * getVestedFraction(beneficiary, block.timestamp)) /
      fractionDenominator;
```
Where:
 - `record.total` is a `uint120` variable
 - getVestedFraction(…) is always <= fractionDenominator

Therefore the claimable amount is always ever `<= record.total` and therefore cannot be higher than the uint120 max.

**Comment by hrishibhat** [comment]
@CholakovAudits Do you have additional comments for the Lead Judge's comment above?

**Comment by cholakovvv** [comment]
@hrishibhat Yes, I have. In all of the reports, I think mine is a duplicate. The last part of the Vulnerability Detail section talks about the downcasting that I reported, so I believe that mine can be validated. Otherwise, the reports I mentioned cannot be totally valid.

**Comment by hrishibhat** [comment]
Result:
Low
Unique
I agree with the Lead judge's comments here:
https://github.com/sherlock-audit/2023-06-tokensoft-judging/issues/70#issuecomment-1668460135
This issue does not mention `initializeDistributionRecord` and is not a duplicate of other issues and not a valid as the value cannot be over uint120 max.

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [CholakovAudits](https://github.com/sherlock-audit/2023-06-tokensoft-judging/issues/70/#issuecomment-1666913696): rejected
